src/main/java/org/javamodularity/moduleplugin/extensions/ModularityExtensionImpl.java

```diff
@@ -0,0 +1,97 @@
+package org.javamodularity.moduleplugin.extensions;
+
+import org.gradle.api.JavaVersion;
+import org.gradle.api.Project;
+import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.api.tasks.compile.JavaCompile;
+import org.javamodularity.moduleplugin.JavaProjectHelper;
+
+import java.util.List;
+
+public class ModularityExtensionImpl implements ModularityExtension {
+
+    private final Project project;
+
+    public ModularityExtensionImpl(Project project) {
+        this.project = project;
+    }
+
+    //region JAVA RELEASE
+
+    //region STANDARD JAVA RELEASE
+    @Override
+    public void standardJavaRelease(int mainJavaRelease) {
+        if (mainJavaRelease < 9) {
+            throw new IllegalArgumentException(String.format(
+                    "Invalid main --release value: %d. Use 'mixedJavaRelease' instead.", mainJavaRelease
+            ));
+        }
+        project.afterEvaluate(p -> configureStandardJavaRelease(mainJavaRelease));
+    }
+
+    private void configureStandardJavaRelease(int mainJavaRelease) {
+        JavaCompile compileJava = helper().compileJavaTask(JavaPlugin.COMPILE_JAVA_TASK_NAME);
+        setJavaRelease(compileJava, mainJavaRelease);
+    }
+    //endregion
+
+    //region MIXED JAVA RELEASE
+    @Override
+    public void mixedJavaRelease(int mainJavaRelease, int moduleInfoJavaRelease) {
+        validateMixedJavaReleaseArgs(mainJavaRelease, moduleInfoJavaRelease);
+
+        CompileModuleOptions moduleOptions = helper().compileJavaTask(JavaPlugin.COMPILE_JAVA_TASK_NAME)
+                .getExtensions().getByType(CompileModuleOptions.class);
+        moduleOptions.setCompileModuleInfoSeparately(true);
+
+        project.afterEvaluate(p -> configureMixedJavaRelease(mainJavaRelease, moduleInfoJavaRelease));
```

Comment:
I'm a little confused about the details, but isn't using lambdas a problem for Gradle to use it's caches? You commented in some other places referencing to this issue. Doesn't the same apply here?

Reply:
Like you said, using lambdas is a problem for Gradle to use its caches. But Gradle doesn't cache `Project.afterEvaluate`; it [caches task inputs](https://github.com/gradle/gradle/pull/6503) so this problem is restricted primarily to `Task.doFirst` and `Task.doLast`. I added a clarifying comment in https://github.com/java9-modularity/gradle-modules-plugin/issues/54#issuecomment-478462010.

Still, I got one of these wrong - it's [`Distribution.contents`](https://github.com/gradle/gradle/blob/e975872ae2cf1d521b73f1107cae772ca09c63cd/subprojects/plugins/src/main/java/org/gradle/api/distribution/internal/DefaultDistribution.java#L51). It can take a lambda (it executes it immediately), but I marked it with a comment as if it couldn't. I'll convert this occurrence to a lambda 🙂

Reply:
Excellent, thanks!